### DISCUSSION TITLE
Moose interprets all grid points with the information from the first point in the CSV

### DISCUSSION THREAD

**Discussion by aaaaaaqing** [idaholab/moose#25022] [Q&A Getting Started] [answered]
I encountered an issue when using PiecewiseConstantFromCSV and PropertyReadFile to read doping information from a CSV file. The problem is that the software seems to only read the data of the first point in the CSV file, meaning that Moose interprets all grid points with the information from the first point in the CSV.

Firstly, let me introduce the format of the CSV file that I want to read. The first two columns represent the X and Y coordinates, respectively, and the third column contains the doping information.

Secondly, let me explain how I used PiecewiseConstantFromCSV and PropertyReadFile in the input file: [Provide the relevant configuration and settings in the input file].
```
[UserObjects]
  [reader_node_n]
  type = PropertyReadFile
  prop_file_name = 'rec_n2_3.csv'
  read_type = 'voronoi'
  nprop = 2
  nvoronoi = 4152 # My CSV file has a total of 4152 points
  []
[]

[Functions]
  [node_n]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = 'reader_node_n'
    read_type = 'voronoi'
    column_number = '2'
  []
[]
```
I am confused as to why such an error occurs, I would appreciate your help

**Comment by GiudGiud** [comment]
I've seen issues with this object when the CSV file has windows line endings. The file is not properly parsed anymore. 

You can use to clean the file:

dos2unix file_name.csv
or
sed -e 's/\r//g' file_name.csv &> file_name_clean.csv

**Reply by aaaaaaqing** [reply]
First of all, thank you for your answer. You mean that there is a problem with my CSV file format, because moose cannot work correctly because of the end of Windows. However, after I used 'dos2unix file_name.csv' to correct my csv file, the problem still hasn't been solved, is there a problem elsewhere

**Reply by GiudGiud** [reply]
Do you mind pasting a few lines from the CSV file?
Is this a 3D simulation with 2D points in the CSV file?

**Reply by aaaaaaqing** [reply]
The sections of my csv file are as follows：
0 0 1.52E+09
1.00E-06 0 1.51E+09
1.00E-06 1.00E-06 1.64E+09
The first column is the coordinates of x, the second column is the coordinates of y, and the third column is the doping information.When I use 'PiecewiseConstantFromCSV' and PropertyReadFile 'to read, all of the dot doping is' 1.52E+09'.

**Reply by aaaaaaqing** [reply]
2D文件

**Reply by aaaaaaqing** [reply]
I am using
 '  for( int a = 1; a < 5; a = a + 1 )
  {
    cout << "p 的值：" << _node_n.value(a) << endl;
    cout << "i 的值：" << a << endl;
  } ' 
  to check whether the doping information is correctly read

**Reply by GiudGiud** [reply]
Can you use Exodus output instead to check? I d rely on that

**Reply by aaaaaaqing** [reply]
Are '.e 'files output by
' [Outputs]
  exodus = true
[]‘viewed through paraview？
Does this'Output 'need any additional modification

**Reply by aaaaaaqing** [reply]
I can't view the data set of the csv file in '.e' through 'paraview'

**Reply by aaaaaaqing** [reply]
I will not get an 'Exodus output' containing csv data, what do I do

**Reply by GiudGiud** [reply]
You wont get a csv from exodus output, but you get a visual representation of the fields and you  can select a certain location in paraview to look at the value there.

You ll need to output to a variable, use a FunctionAux auxkernel for that

**Reply by aaaaaaqing** [reply]
The result from the auxiliary kernel is still only the first value, and that should not be the problem

**Reply by GiudGiud** [reply]
Ok thanks for checking

I m worried the 3rd column is being interpreted as the Z coordinate. Can you please add a column of zeros there

**Reply by aaaaaaqing** [reply]
Still to no avail, the output is still all the values of the first point, I wonder if it is because of the wrong use of '_node_n.value(_qp)' to read the value of each point in the function

**Reply by GiudGiud** [reply]
where did you see this?
_node_n.value(_qp) ??

what are the dimensions of your mesh? You're aware that voronoi is trying to do a "nearest point" algorithm to see the values right?

**Reply by aaaaaaqing** [reply]
I just used the _ node_n.value (_ qp) to read the value of the function. For example, in the auxiliary kernel, I used return _ node_n.value (_ qp) to assign the value of the function to the auxiliary variable.I use the grid size is 1e-7m or so, is it because the grid size is too small to cause nearest point can not read correctly, but I use node mode and modify the corresponding csv file format can not be correctly read, there is still only the first point of the value.

**Reply by GiudGiud** [reply, marked as answer]
yeah that's not going to work
value() on a function takes the time, not the quadrature point.

You need to do this
```
function.value(_t, _q_point[_qp])
```
with the time and the current quadrature point

**Reply by aaaaaaqing** [reply]
When I use your method to type, there is an error:
error: no match for call to '(const MooseArray<libMesh::Point>) (unsigned int&)'
   28 |   return _node_n.value(_t,_q_point(_qp));
So what's going on here.

**Reply by GiudGiud** [reply]
You must use square brackets not parentheses

**Reply by aaaaaaqing** [reply]
When I woke up, I found that the results of moose simulation have all changed, although there was always a point value before, but today when I carried out a more bizarre thing, all the values are 0, including the variables in the kernel, I simply can not understand what happened

**Reply by GiudGiud** [reply]
You probably saved before going to sleep.
You can use ctrl+z to see what you changed last

**Reply by aaaaaaqing** [reply]
The computer shut down and I couldn't use 'ctrl +z' to undo

**Reply by aaaaaaqing** [reply]
I modified the problem of all zeros and tried the 'function.value(_t, _q_point[_qp])' you mentioned. The simulation results did change, but the simulation results were still incorrect. When using：
’Real
test::computeValue()
{
  return _node_n.value(_t,_q_point[_qp]);
}‘
to get the value of the auxiliary kernel, the result is still the value of only one point (but no longer the value of the first point).
I wonder if this is the only way to read the CSV file, and whether I should read the CSV file into the material information and then use it is the correct way?

**Reply by GiudGiud** [reply]
Did you add the third column with all zeros yet?

**Reply by aaaaaaqing** [reply]
When I use '_node_n.value (_t, _q_point[_qp])', the resulting image of the simulation has appeared in general outline, I believe this is the correct way, although the auxiliary kernel still displays a value, but I believe this is irrelevant, thank you for your continued help, thank you
